packages/kit/test/apps/basics/test/cross-platform/client.test.js

```diff
@@ -221,15 +221,24 @@ test.describe('Navigation lifecycle functions', () => {
 	});
 
 	test('beforeNavigate is triggered after clicking a download link', async ({ page, baseURL }) => {
-		await page.goto('/navigation-lifecycle/before-navigate/prevent-navigation');
+		const pathname = '/navigation-lifecycle/before-navigate/prevent-navigation';
 
-		await page.click('a[download]');
-		expect(await page.innerHTML('pre')).toBe('0 false undefined');
+		await page.goto(pathname);
+
+		const download = page.waitForEvent('download', { timeout: 3000 });
+		await page.locator('a[download]').click();
+		await (await download).cancel();
+
+		const current_url = baseURL + pathname;
+
+		expect(page.url()).toBe(current_url);
+		expect(await page.locator('pre').innerText()).toBe('0 false undefined');
 
 		await page.click('a[href="/navigation-lifecycle/before-navigate/a"]');
+		await page.waitForLoadState('networkidle');
```

Comment:
if this is just a navigation, could we use the existing util for this?
```suggestion
		await clicknav('a[href="/navigation-lifecycle/before-navigate/a"]');
```

Reply:
You're right we can use it. I wasn't familiar with the existing utils. Should we always use those as much as possible?

Reply:
yes, we should always use `clicknav` if navigating to another page. In a normal webapp, playwright can tell that the navigation is complete. But when you do a client-side navigation, it doesn't know when navigation is complete, so we had to write some code to notify when the page load is done and use that

Reply:
I tried it but realised it wasn't applicable in this case since navigation is always prevented on this route.